### ISSUE TITLE
exp(p3): potential-based team-welfare shaping — Ng-Harada-Russell aligned shaping

### ISSUE THREAD

**Issue by rjwalters** [rjwalters/bucket-brigade#283]
## Thesis context

Action-shaping (#262, tier-3) added per-step bonuses but didn't shift the verdict. The thesis suggests *non-aligned* shaping won't help; we need shaping that's mathematically aligned with the long-horizon objective.

**Ng, Harada & Russell (1999)** proved that potential-based shaping `r' = r + γΦ(s') − Φ(s)` leaves the optimal policy invariant while changing learning dynamics. If we set `Φ(s) = team_welfare(s)` — e.g., the heuristic specialist's estimated value at state s — we get a per-step reward stream that's locally informative about the cooperative goal without distorting the policy class.

## Protocol

1. Compute Φ(s) from specialist policy: roll out specialist many times, fit a Φ regressor `state → mean team return from this state`
2. Add `γΦ(s') − Φ(s)` as a per-step shaping bonus on top of existing rewards
3. Train PPO with this shaped reward; compare gap_closed vs the 0.075-0.18 baseline range
4. Verify policy invariance theoretically and (cheaply) empirically — does the optimal policy still concentrate on specialist-like behavior?

## Why this differs from action-shaping (#262)

#262 added shaping that's NOT potential-based — it doesn't satisfy NHR's invariance property. The shaped optimal policy may differ from the original. Potential-based shaping is principled and invariant.

## Verdict logic

| Outcome | Interpretation |
|---|---|
| gap_closed ≥ 0.5 | Aligned shaping was missing. Per-step info just needed to be aligned, not absent. |
| 0.25-0.5 | Partial; layer with other interventions. |
| <0.25 | Shaping isn't enough; need algorithmic change (COMA / LOLA). |

## Effort

~1 day implementation; ~1h smoke.

## Sequencing

Run AFTER #270. Especially relevant if anti-attractor verdict.

## References

- Ng, Harada, Russell (1999) "Policy invariance under reward transformations"
- `research_notebook/2026-05-17_thesis_misaligned_gradients.md`
- PR #268 (action-shaping tier-3 — what this builds on / fixes)

---

### Status note

Per @rjwalters's 2026-05-17 comment, the *anti-attractor* framing for this experiment was **falsified by #271** (random-init MLP best-of-N comparable to PPO — no active repulsion from the specialist region). This issue **stays filed as a defensible long-shot** for *basin-locality*: a state-value-aligned shaping signal may pull policies toward higher-V states more effectively than the local advantage gradient, even absent an anti-attractor. Priority drops from "Phase-2-load-bearing" to "long-shot worth running if cheap." Re-validate value if a higher-priority intervention reopens the basin question.

### Problem statement

After five PPO-side interventions (MAPPO #225/#231, obs-fix #220/#233, spatial #221/#228, curriculum #260/#266, action-shaping #259/#262/#268) all returned tier-3 verdicts and #270 (BC-init basin probe) plus #271/#272 falsified anti-attractor, the working thesis (`research_notebook/2026-05-17_thesis_misaligned_gradients.md:72`) is that PPO's local advantage gradient is *misaligned* with the long-horizon cooperative objective on bucket-brigade. Per-step rewards exist but are not informative about whether the current step is making team welfare better or worse. Potential-based shaping per Ng-Harada-Russell (1999) is the canonical mechanism for injecting an aligned per-step signal without distorting the optimal policy.

### Theoretical basis (NHR 1999)

**Theorem** (Ng, Harada, Russell 1999, "Policy Invariance Under Reward Transformations: Theory and Application to Reward Shaping", ICML). For any potential function `Φ: S → R` and any base MDP `M = (S, A, P, R, γ)`, the shaped MDP `M' = (S, A, P, R', γ)` with
```
R'(s, a, s') = R(s, a, s') + γ Φ(s') − Φ(s)
```
has the same set of optimal policies as M. The proof relies on the telescoping sum: across an episode, `Σ_t γ^t (γΦ(s_{t+1}) − Φ(s_t)) = γ^T Φ(s_T) − Φ(s_0)`, which is a constant offset independent of policy (modulo the terminal-state convention, where Φ(terminal) must be set to 0 to preserve invariance exactly). Thus the optimal Q-function shifts by Φ(s) but argmax is preserved.

**What we get**: a per-step informative signal (the advance/regress of Φ) overlaid on the same equilibrium structure. Faster convergence is *not* guaranteed in general, but is the empirically expected outcome when Φ is a good proxy for V*.

### Candidate potential functions Φ(s)

| # | Φ(s) | Pros | Cons |
|---|------|------|------|
| **A** | **Specialist-rollout Monte-Carlo regressor**: fit `Φ_θ(s) ≈ E_{π_specialist}[Σ_t γ^t r_t \| s_0=s]` on ~10k specialist rollouts (reuse `bc_init.py` data path) | Cheap; uses existing specialist rollouts; aligned with cooperative target | Approximates V^{π_specialist}, not V*; risk of nudging policy *toward* specialist but not *past* it |
| **B** | **Closed-form team-welfare**: Φ(s) = w₁·(num SAFE houses) − w₂·(num RUINED houses) − w₃·(num BURNING houses); coefficients matched to scenario's team-reward weights | Zero training cost; interpretable; trivially debuggable | Coarse; doesn't credit progress within state types |
| **C** | **Burning-house decay potential**: Φ(s) = −Σ_{burning h} expected_time_to_ruin(h); penalizes states with un-attended fires | Captures urgency; matches save-the-house objective | Requires modeling expected burnout dynamics |

**Recommended starting choice**: **B** (closed-form, zero learning) for the smoke test, then **A** (specialist-MC) for the headline run. **A** matches the original protocol; **B** is a cheaper falsification path that exposes whether the issue is "needs *any* aligned signal" vs "needs a *specialist-quality* aligned signal."

### Reward integration point

Shaping must be added inside the per-step reward computation, after existing components are computed:

- **Python path**: `bucket_brigade/envs/bucket_brigade_env.py:294` (`_compute_rewards`). Add `Φ(s')` term using `self.houses` (current) and `prev_houses` (cached at line 297) to compute `γ·Φ(s') − Φ(s)`. Mirror the action-shaping plumbing at lines 408–444: gate behind a scenario field (e.g., `potential_shaping_lambda: float = 0.0`) so existing scenarios are byte-identical when unused.
- **Rust path**: `bucket-brigade-core/src/engine/rewards.rs:5` (`compute_rewards`). Mirror Python implementation to keep parity (the Rust path is the perf-critical one used by training).
- **Scenario field**: add to `bucket_brigade/envs/scenarios_generated.py:19` (`Scenario` dataclass) — follow the pattern at line 99 (`action_shaping_alpha`). Also add to `bucket-brigade-core/src/scenarios.rs` per the generated-file comment at line 10.
- **Trainer wiring**: pipe through `experiments/p3_specialization/train.py:625` next to the existing `action_shaping_alpha` mutation. For potential type **A**, also accept a `--phi-checkpoint` arg pointing at the trained Φ regressor; load lazily inside `env_fn` (line 628).

### Terminal-state convention

NHR's invariance requires Φ(terminal) = 0. `bucket_brigade_env.py:148–218` returns `dones` as the truncation flag; at the truncated last step, compute the shaping term as `0 − Φ(s)` (i.e., treat next-state potential as 0). Verify by inspecting `_check_termination` at line 448.

### Φ(s) regressor design (option A)

- **Architecture**: 2-layer MLP, hidden 64, ReLU. Input: env observation flattened (use `_flatten_per_agent` from `joint_trainer.py:413`). Output: scalar V estimate.
- **Training data**: collect ~10k specialist rollouts on `minimal_specialization` (reuse `bucket_brigade/baselines/specialist.py:140` `SpecialistAgent`). For each `(s_t, return_t)` pair where `return_t = Σ_{k≥t} γ^k r_k`, fit MSE.
- **Validation**: hold-out 1k rollouts; report `R²` of Φ vs Monte-Carlo return. Reject if R² < 0.5 (Φ isn't fitting; shaping signal will be noise).
- **Suggested location**: `experiments/p3_specialization/fit_phi.py`. Saves `phi.pt` next to BC checkpoints.

### Acceptance criteria

- [ ] `Scenario.potential_shaping_lambda` (default 0.0) and `Scenario.potential_shaping_kind` (default `"none"`) fields added in both Python (`scenarios_generated.py`) and Rust (`scenarios.rs`). Pre-#283 scenarios are byte-identical (`lambda == 0.0`).
- [ ] Φ implementations: at least option **B** (closed-form team-welfare) wired up; option **A** (specialist-MC regressor) wired up if **B** doesn't move the needle.
- [ ] Shaped reward `F = γ·Φ(s') − Φ(s)` added to `_compute_rewards` in both Python and Rust, gated by lambda. Terminal-state convention enforced (Φ(terminal)=0).
- [ ] Smoke unit test (`tests/test_environment.py`): policy invariance check — verify that for a fixed random policy, total *discounted episode return* under shaped reward equals base return + `(γ^T·Φ(s_T) − Φ(s_0))` to within float tolerance, confirming the telescoping identity.
- [ ] Training run on `minimal_specialization`: 50 PPO iterations, 3 seeds, lambda sweep `{0.0, 0.5, 1.0, 2.0}`. Compare `gap_closed` vs the 0.075–0.18 random-init PPO baseline (`MINSPEC_RANDOM = −96.07`, `MINSPEC_SPECIALIST = −22.07`; `bc_init.py:52-53`).
- [ ] Notebook verdict (`research_notebook/2026-05-17_issue283_potential_shaping_verdict.md`) using the gap_closed thresholds from the original issue.

### Success criteria

- **Primary**: `gap_closed ≥ 0.5` at any `lambda > 0` setting on `minimal_specialization`, with the lambda=0 control reproducing the 0.075–0.18 random-init baseline. This would validate aligned per-step signal as a missing ingredient.
- **Secondary (invariance check)**: under deterministic-argmax eval, the *unshaped* mean episode return at convergence under `lambda > 0` should be statistically indistinguishable from a `lambda = 0` run that has converged. (NHR guarantees same optimum; this verifies the implementation is correct, not just better.)
- **Negative**: `gap_closed < 0.25` for all lambda → escalates to algorithmic interventions (HCA #289, Social Influence #290, or COMA/LOLA per #193).

### Affected Files

- `bucket_brigade/envs/bucket_brigade_env.py` — extend `_compute_rewards` (line 294) with shaping term; add Φ accessor.
- `bucket_brigade/envs/scenarios_generated.py` — add `potential_shaping_lambda`, `potential_shaping_kind` fields to `Scenario` (line 19); update `__post_init__`.
- `bucket-brigade-core/src/engine/rewards.rs` — mirror Python shaping in `compute_rewards` (line 5).
- `bucket-brigade-core/src/scenarios.rs` — mirror new scenario fields.
- `definitions/scenarios.json` — no edits needed (defaults preserve existing behavior); optional new scenario `minimal_specialization_phi_a` for clarity.
- `experiments/p3_specialization/fit_phi.py` (NEW) — specialist-MC Φ regressor training script (option A).
- `experiments/p3_specialization/train.py` — wire `potential_shaping_lambda`, `potential_shaping_kind`, `--phi-checkpoint` through `CellConfig` (line 58) and the mutation block at line 625.
- `tests/test_environment.py` — telescoping-identity invariance test.
- `research_notebook/2026-05-17_issue283_potential_shaping_verdict.md` (NEW) — final writeup.

### Test Plan

- [ ] **Unit (invariance)**: in `tests/test_environment.py`, fix a seed, roll out a random policy for one episode under `lambda=0` (base) and `lambda=1.0` (shaped, option B). Assert `sum(shaped_rewards * γ^t) ≈ sum(base_rewards * γ^t) + γ^T·Φ(s_T) − Φ(s_0)` to within `1e-5` per agent. Use the closed-form Φ from option **B** so the test has no learned-Φ dependency.
- [ ] **Unit (terminal)**: assert that at truncated last step, the shaping term equals `−Φ(s)` (Φ(next)=0 enforced).
- [ ] **Unit (zero-lambda)**: assert that with `lambda=0` the per-step rewards are byte-identical to pre-#283 (regression guard).
- [ ] **Integration (smoke)**: 1-hour smoke run, `minimal_specialization`, `lambda=1.0`, option B, 1 seed, 10 PPO iters. Verify training loop runs end-to-end and `metrics.json` records shaped + unshaped reward streams separately for diagnosis.
- [ ] **Main experiment**: on `COMPUTE_HOST_PRIMARY` per `experiments/REMOTE_EXECUTION.md`, lambda sweep `{0.0, 0.5, 1.0, 2.0}` × option `{B, A}` × 3 seeds × 50 iters on `minimal_specialization`. Estimated cost ~1 day wall-clock.
- [ ] **Invariance empirical check**: at the best lambda, take the converged policy and evaluate it under the *unshaped* reward — confirm score matches the lambda=0 run's converged policy (within seed noise). Detects implementation bugs where Φ accidentally distorts the policy.
- [ ] **Verdict notebook**: writeup with gap_closed table, lambda sensitivity plot, invariance check, and verdict per the original thresholds.

### Dependencies

(none — #270 already merged via #278; all prerequisites resolved)

### Related work / cross-references

- **#193** — Parent proposal listing reward-shaping interventions for PPO learnability.
- **#259 / #262 / #268** — Action-conditioned shaping (the non-potential-based shaping this builds on / corrects); confirmed flat → motivates trying *aligned* shaping.
- **#265** — Dense progress signal (intervention #4 from #193). Sibling shaping intervention; **potential overlap risk** — coordinate Φ choice so this experiment isn't subsumed.
- **#270 / #271 / #272** — Basin-trap probes; #271 falsifies anti-attractor framing → re-prioritization noted above.
- **#289 (HCA)**, **#290 (Social Influence)** — Adjacent algorithmic interventions; escalate to these if #283 verdict is negative.
- **#234** — Rules-level redesign (env v2). Orthogonal but listed for completeness; if #283 is negative *and* #289/#290 are negative, the env may need redesign.
- `research_notebook/2026-05-17_thesis_misaligned_gradients.md:72` — Working thesis citing NHR as untried.
- `research_notebook/2026-05-17_lit_review.md:148` — Lit review noting potential-based team-welfare shaping is the canonical untried aligned-shaping option.

### Compute & ops

- **NEVER run training locally** (see `CLAUDE.md` § COMPUTE RESOURCE GUIDELINES). Use `COMPUTE_HOST_PRIMARY` via `source .env && ssh $COMPUTE_HOST_PRIMARY` and run inside `tmux`. Smoke can run locally (single seed, 10 iters).
- **Φ regressor (option A) training**: 10k specialist rollouts + 1k SGD steps fits a 2-layer MLP in minutes on CPU; can run anywhere.

**Comment by rjwalters** [comment]
## Re-prioritization post-#271

Potential-based shaping (Ng-Harada-Russell) was filed as a fix for the active-anti-attractor reading — aligned per-step rewards that don't distort optimal policy. With anti-attractor falsified by #271, the urgency drops.

Potential-based shaping might still help with basin-locality (a state-value-aligned shaping signal could pull policies toward higher-V states more effectively than pure local gradient), but it's no longer Phase-2-load-bearing. Stays filed as a defensible long-shot.

**Comment by rjwalters** [comment]
Curating this issue per user request

**Comment by rjwalters** [comment]
Curation complete (2026-05-17). Body amended with:

- NHR (1999) theorem statement with telescoping-sum justification and terminal-state convention requirement (Φ(terminal)=0).
- Three candidate Φ(s) functions (closed-form team-welfare, specialist Monte-Carlo regressor, burning-house decay) with pros/cons table; recommended starting choice **B** for falsification, **A** for the headline run.
- Concrete reward-integration points with file:line refs: Python `bucket_brigade_env.py:294`, Rust `rewards.rs:5`, scenario field next to `action_shaping_alpha` (`scenarios_generated.py:99`), trainer wiring at `train.py:625`. Mirrors the action-shaping (#259/#262) plumbing pattern.
- Acceptance criteria, Affected Files, Test Plan (includes the telescoping-identity invariance unit test that distinguishes a correct NHR implementation from a buggy one), and Dependencies sections.
- Cross-references to #193 (parent), #259/#262/#268 (predecessor non-aligned shaping), #265 (sibling dense-progress, overlap risk flagged), #270/#271/#272 (basin probes), #289/#290 (escalation if negative), and the working thesis + lit review.
- Updated priority framing per the original 2026-05-17 status comment: long-shot for basin-locality after anti-attractor was falsified by #271. Awaits human approval (`loom:issue`) before Builder claim.
